pkg/kubelet/cm/dra/claiminfo.go

```diff
@@ -166,24 +163,33 @@ func newClaimInfoCache(stateDir, checkpointName string) (*claimInfoCache, error)
 				return nil, fmt.Errorf("failed to add CDIDevices to claimInfo %+v: %+v", info, err)
 			}
 		}
-		cache.add(info)
+		cache.addOrUpdate(&info)
 	}
 
 	return cache, nil
 }
 
-func (cache *claimInfoCache) add(res *ClaimInfo) {
+func (cache *claimInfoCache) addOrUpdate(res *ClaimInfo) {
 	cache.Lock()
 	defer cache.Unlock()
 
 	cache.claimInfo[res.ClaimName+res.Namespace] = res
 }
```

Comment:
/hold

I think we need to update this to *actually* do an update if there is already an entry in the cache (rather than simply overwriting it). Otherwise, previously updated podReferences will be overwritten, for example.

Reply:
@klueska I've added `addPodReference` and `deletePodReference` APIs to the `claimInfoCache` to do in-place updates. There is still one place in the `PrepareResources` code where I assumed that replacing ClaimInfo object is safe. Not sure you'll like it, so it's just a proposal. Hopefully you'll have an idea how to do it better.

Reply:
@klueska I'm considering making `addCDIDevices` API to the `claimInfoCache` as well as this is a last one in the `ClaimInfo` that updates its object. Do you think it makes sense?

Reply:
> Please add unit tests for all of this code...

The code is covered by more than 80%. I'll look if some of the new code is not covered and add some tests if this is the case. Thanks for reminding.

Reply:
In fact the only thing that should be done on top of this is updating CDI devices after successful `PrepareResources` GRPC call instead of replacing the whole claim. I'll do it today.

Reply:
> The code is covered by more than 80%. I'll look if some of the new code is not covered and add some tests if this is the case.

What triggered my "add unit tests" comment was that there was a "addOrUpdate" method but apparently no unit test for the "update" usage of that method. This is a perfect example why the code coverage percentage is not a good indicator for how well something is tested: this method would have been covered 100% with a single, simple test, but that would have ignored all edge cases.

Reply:
@klueska I've updated the code again to make it more clear and straightforward. Now all updates of cached `ClaimInfo` objects are done through `claimInfoCache` APIs, which lock manager cache. PTAL.

Reply:
claimInfoCache doesn't seem to have any unit tests? Testing it indirectly may trigger bugs, but isn't the same as writing dedicated unit tests that can be reviewed to ensure that edge cases are exercised.

At the higher level, https://github.com/kubernetes/kubernetes/blob/9ca60a21fdde35ab01d08827b170759b032921ba/test/e2e_node/dra_test.go#L69-L133 should cover various scenarios for "NodeUnprepareResources must be called":
- kubelet restarts while several unrelated claims are prepared for different pods
- kubelet restarts while several pods use the same claim

Reply:
I agree with patrick -- we need standalone unit tests agains the operations on the `claimInfoCache` type itself.

Reply:
Sure, I'll add unit tests for the claiminfo.go.